### PR TITLE
v2dl3-vegas: Event Selection Patch

### PR DESCRIPTION
**Pull request by tmcahill**
This patch utilizes the previously implemented "EventClass" object in order to select events to separate event lists and FITS files.
New flags have been tested and output has been reproduced with no differences to main using fitsdiff.

Very little VEGAS-specific review here, so I've added a question about how we utilize VEGAS root data in the previous patch.

## For VEGAS Review
 - `fillEVENTS_not_safe.py` accesses additional `selectedEventsTree` attributes `.fMSW` and `.fMSL` in the manner specified by VEGAS docs
 - From previous commit 6c6e3f2: 
    - `EventClass.py` reads EA via VARootIO to get the cuts info ("pfCutsFileText" in the root file)
    - `fCutsFileText` is passed (as a string) to `getCuts()` in `util.py` to search for substrings matching the desired cut param names.
       - This seems to me to be a naive method, but I could not find info about this in the VEGAS docs. I did a fair amount of investigating the EA root files via VARootIO and uproot; it seems to me that cuts info is only available as raw text. However, if there is a VARootIO library method that parses this file for me, that is better to use so it will be paired to VEGAS updates.
       
## Full Changelog
`fillEVENTS.py`
 - Event class support when provided optional arg `event_class_idx`
 - Appends MSW and MSL columns when appropriate
 
 `fillRESPONSE.py`
 - Event class support when provided optional arg `event_class_idx`
 
 `genHDUList.py`
 - Optional arguments added
 
 `v2dl3_for_vegas.py`
 - `-save_msw_msl` flag added
 - Extended fits output for event class mode as new helper function `make_eclass_path()`
 - Extended index file output for event class mode as new helper function `gen_index_files()`
 
 `VegasDataSource.py`
 - Support for event classes and `save_msw_msl` flag
 
 `fillEVENTS_not_safe.py`
 - Now filters events to separate dicts when using >1 event class.
 - Added MSW and MSL access from SelectedEventsTree

### CODE REVIEW

File: pyV2DL3/script/v2dl3_for_vegas.py [hunk 1]
```diff
@@ -202,26 +213,94 @@ def cli(
...
-                logging.info(
-                    f"Overwriting OBS_ID={hdulist[1].header['OBS_ID']} with OBS_ID={fname_base}"
-                )
-                hdulist[1].header["OBS_ID"] = fname_base
-            hdulist.writeto(f"{output}/{fname_base}.fits", overwrite=True)
-            flist.append(f"{output}/{fname_base}.fits")
-            # Generate hdu obs index file
+
+            # Prepare output paths
+            output_path = f"{output}/{fname_base}"
```
Comment (GernotMaier) [inline pyV2DL3/script/v2dl3_for_vegas.py:224]: Super minor comment: `os.path.join()` provides a save way of joining paths and files and is sometimes preferred. I am just mention it, probably more a matter of style.

**Review by GernotMaier — APPROVED** [review]
Two very minor comments, otherwise all good!

File: pyV2DL3/script/v2dl3_for_vegas.py [hunk 2]
```diff
@@ -202,26 +213,94 @@ def cli(
...
+def gen_index_files(flist, output, eclass_count=1):
+    # Generate master index files
+    logging.info(
+        f"Generating index files {output}/obs-index.fits.gz "
+        f"and {output}/hdu-index.fits.gz"
+    )
+    create_obs_hdu_index_file(flist, output)
+
+    # Generate index files per event class
+    if eclass_count > 1:
```
[the remark this answers is not in the document]
Comment (tmcahill) [reply pyV2DL3/script/v2dl3_for_vegas.py:267]: Good question. Internally, we are still using one "event class" when not in `event_class_mode`. Perhaps I should rename those variables to "event_list" when not referring to an EventClass object.

The reason it is necessary is that the output directories are split into 'ec' subdirectories only when there is more than one event class. In that case, it is helpful for the user to have both a set of master index files as well as index files for each 'ec' subdirectory. When not using `event_class_mode`, `eclass_count` will be == 1, but we don't want that block to run in that case, as the above block has already created the index files.

**Review by cameronrulten — COMMENTED** [review]
Everything seems sensible. My only comment is a minor one about style when constructing strings. For consistency I would stick to using f-strings through out instead of the mix and match approach.
Apart from this really minor detail, I presume you have tested these changes and everything is working as intended?

**Comment by tmcahill** [comment]
@cameronrulten @GernotMaier 
All paths now use `os.path` operations. I also renamed internal `event_class` variables to `event_group` related names to be more clear whether or not it is referring specifically to an EventClass object. Changes are re-tested and working

@cameronrulten 
Good question on the testing. I do test all of the pull requests, and our demo branch where these features are drafted has also been used and tested extensively. 
Right now, the tests are done by hand, unfortunately. I re-run a number of different flag combinations to compare to main using fitsdiff, that way all differences (if any) in output are accounted for. We have a plan to set up CI for automated v2dl3-vegas testing soon.

**Comment by GernotMaier** [comment]
All fine with me, thanks!

**Pull request merged**